x-pack/metricbeat/module/aws/ec2/ec2.go

```diff
@@ -0,0 +1,208 @@
+package ec2
+
+import (
+	"fmt"
+	"time"
+
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/defaults"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/aws"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("aws", "ec2cpu", New,
+		mb.DefaultMetricSet(),
+	)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	*aws.MetricSet
+	config *aws.Config
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Experimental("The aws ec2cpu metricset is experimental.")
+
+	config := aws.Config{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		config:    &config,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	sess, err := session.NewSession()
+	if err != nil {
+		report.Error(errors.Wrap(err, "Error creating new session"))
+	}
+
+	creds := credentials.NewChainCredentials(
```

Comment:
I prefer this goes to init function but depends on authentication method and how long the temporary creds will last, this might needed more frequent than just once in init function. I'm planning to move this after I figure out what kind of authentication will be used.